### ISSUE TITLE
ONNX cannot save the XGBoost binary classifier properly when trained on an imbalanced dataset.

### ISSUE THREAD

**Issue by cugurm** [pytorch/pytorch#150805]
### 🐛 Describe the bug

ONNX cannot properly save an XGBoost binary classification model when it is trained on an imbalanced dataset.

When I create the dataset for the XGBoost binary classification model like this:
```
    n_instances, n_features = 100_000, 300
    X = np.random.rand(n_instances, n_features)  
    y = np.random.randint(0, 2, size=(n_instances,))  # Binary labels (0 or 1)
```
I am able to save the trained model to ONNX, load it, and make predictions that match the original model.

However, when I create the training dataset with imbalanced labels like this:
```
    n_instances, n_features = 100_000, 300
    X = np.random.rand(n_instances, n_features)  
    class_0_count, class_1_count = 90_000, 10_000 
    y = np.concatenate([np.zeros(class_0_count), np.ones(class_1_count)])
    np.random.shuffle(y)
```
saving the model to ONNX and loading it results in predictions that differ from the original model.

Reproducer:
```
import numpy as np
import onnxruntime as rt
from sklearn.datasets import load_iris
from xgboost import XGBClassifier
from skl2onnx import convert_sklearn
from skl2onnx.common.data_types import FloatTensorType
from skl2onnx import update_registered_converter
from skl2onnx.common.shape_calculator import calculate_linear_classifier_output_shapes
from onnxmltools.convert.xgboost.operator_converters.XGBoost import convert_xgboost

def convert_xgboost_pipeline_to_onnx(X, y, n_features, n_test_instances=3):
    # Train XGBoost classifier
    pipe = XGBClassifier(objective='binary:logistic', eval_metric='logloss', n_estimators=500, max_depth=5, reg_lambda=1, reg_alpha=0)
    pipe.fit(X, y)

    # Register the ONNX converter for XGBClassifier
    update_registered_converter(
        XGBClassifier,
        "XGBoostXGBClassifier",
        calculate_linear_classifier_output_shapes,
        convert_xgboost,
        options={"nocl": [True, False], "zipmap": [True, False, "columns"]},
    )

    # Convert the model to ONNX
    model_onnx = convert_sklearn(
        pipe,
        "pipeline_xgboost",
        [("input", FloatTensorType([None, n_features]))],
        target_opset={"": 12, "ai.onnx.ml": 2},
        options={"zipmap": False, "nocl": False},
    )

    # Save the ONNX model
    with open("pipeline_xgboost.onnx", "wb") as f:
        f.write(model_onnx.SerializeToString())

    # Compare predictions
    print("XGBoost predict:", pipe.predict(X[:n_test_instances]))
    print("XGBoost predict_proba:", pipe.predict_proba(X[:n_test_instances]))

    # Predictions with ONNX Runtime
    sess = rt.InferenceSession("pipeline_xgboost.onnx", providers=["CPUExecutionProvider"])
    pred_onx = sess.run(None, {"input": X[:n_test_instances].astype(np.float32)})
    print("ONNX predict:", pred_onx[0])
    print("ONNX predict_proba:", pred_onx[1])


if __name__ == "__main__":
    n_instances, n_features = 100_000, 300
    
    X = np.random.rand(n_instances, n_features)  
    y = np.random.randint(0, 2, size=(n_instances,))  # Binary labels (0 or 1)
    convert_xgboost_pipeline_to_onnx(X, y, n_features)  
    
    class_0_count, class_1_count = 90_000, 10_000 
    y = np.concatenate([np.zeros(class_0_count), np.ones(class_1_count)])
    np.random.shuffle(y)
    convert_xgboost_pipeline_to_onnx(X, y, n_features) 
```

### Versions
```
Collecting environment information...
PyTorch version: 2.0.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Feb  4 2025, 14:57:36) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.8.0-52-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA RTX 2000 Ada Generation Laptop GPU
Nvidia driver version: 535.183.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               20
On-line CPU(s) list:                  0-19
Vendor ID:                            GenuineIntel
Model name:                           13th Gen Intel(R) Core(TM) i9-13900H
CPU family:                           6
Model:                                186
Thread(s) per core:                   2
Core(s) per socket:                   14
Socket(s):                            1
Stepping:                             2
CPU max MHz:                          5400.0000
CPU min MHz:                          400.0000
BogoMIPS:                             5990.40
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            544 KiB (14 instances)
L1i cache:                            704 KiB (14 instances)
L2 cache:                             11.5 MiB (8 instances)
L3 cache:                             24 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-19
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Mitigation; Clear Register File
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu11==11.10.3.66
[pip3] nvidia-cublas-cu12==12.1.3.1
[pip3] nvidia-cuda-cupti-cu11==11.7.101
[pip3] nvidia-cuda-cupti-cu12==12.1.105
[pip3] nvidia-cuda-nvrtc-cu11==11.7.99
[pip3] nvidia-cuda-nvrtc-cu12==12.1.105
[pip3] nvidia-cuda-runtime-cu11==11.7.99
[pip3] nvidia-cuda-runtime-cu12==12.1.105
[pip3] nvidia-cudnn-cu11==8.5.0.96
[pip3] nvidia-cudnn-cu12==8.9.2.26
[pip3] nvidia-cufft-cu11==10.9.0.58
[pip3] nvidia-cufft-cu12==11.0.2.54
[pip3] nvidia-curand-cu11==10.2.10.91
[pip3] nvidia-curand-cu12==10.3.2.106
[pip3] nvidia-cusolver-cu11==11.4.0.1
[pip3] nvidia-cusolver-cu12==11.4.5.107
[pip3] nvidia-cusparse-cu11==11.7.4.91
[pip3] nvidia-cusparse-cu12==12.1.0.106
[pip3] nvidia-nccl-cu11==2.14.3
[pip3] nvidia-nccl-cu12==2.19.3
[pip3] nvidia-nvjitlink-cu12==12.4.99
[pip3] nvidia-nvtx-cu11==11.7.91
[pip3] nvidia-nvtx-cu12==12.1.105
[pip3] onnx==1.14.0
[pip3] onnxconverter-common==1.14.0
[pip3] onnxmltools==1.11.2
[pip3] onnxruntime==1.20.0
[pip3] skl2onnx==1.15.0
[pip3] torch==2.0.1
[pip3] torch-geometric==2.3.1
[pip3] torchmetrics==1.1.1
[pip3] triton==2.0.0
[conda] Could not collect
```

**Comment by cugurm** [comment]
Related to ONNX and XGboost not to PyTorch.
